lib/ProducerImpl.cc

```diff
@@ -912,10 +912,9 @@ bool ProducerImpl::ackReceived(uint64_t sequenceId, MessageId& rawMessageId) {
 
     if (op.chunkedMessageId) {
```

Comment:
It seems that `OpSendMsg::chunkedMessageId` field is not necessary. It's added just to indicate if the current `OpSendMsg` is related to a chunk. But `OpSendMsg::numChunks` can also be used for that.

Without saving an empty `ChunkMessageIdImpl`, we can just merge `setChunkedMessageIds` to the current constructor:

```c++
    ChunkMessageIdImpl(std::vector<MessageId>&& chunkedMessageIds)
        : chunkedMessageIds_(std::move(chunkedMessageIds)) {
        auto lastChunkMsgId = chunkedMessageIds_.back();
        this->ledgerId_ = lastChunkMsgId.ledgerId();
        this->entryId_ = lastChunkMsgId.entryId();
        this->partition_ = lastChunkMsgId.partition();
    }
```

After that, in `ConsumerImpl` we can write:

```c++
    ChunkMessageIdImplPtr chunkMsgId =
        std::make_shared<ChunkMessageIdImpl>(chunkedMsgCtx.moveChunkedMessageIds());
```

and in `ProducerImpl`:

```c++
    if (op.numChunks > 1) {
        // Handling the chunk message id.
        op.chunkMessageIdList.push_back(messageId);
        if (op.chunkId == op.numChunks - 1) {
            auto chunkedMessageId = std::make_shared<ChunkMessageIdImpl>(std::move(op.chunkMessageIdList));
            messageId = chunkedMessageId->build();
        }
```